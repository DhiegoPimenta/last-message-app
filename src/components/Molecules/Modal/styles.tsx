import styled, { css } from 'styled-components'

export const Overlay = styled.div`
    ${({ theme }) => css`
        background: ${theme.colors.overlay};
        width: 100vw;
        height: 100vh;
        position: absolute;
        z-index: ${theme.layers.base};
        left: 0;
        top: 0;

        align-items: center;
        justify-content: center;
    `})}
`

export const Wrapper = styled.div`
    ${({ theme }) => css`
        width: 45rem;
        height: 30rem;
        border: 4px solid ${theme.colors.danger};
        border-radius: ${theme.border.radius};
        display: flex;
        flex-direction: column;
        justify-content: space-around;
        z-index: ${theme.layers.alwaysOnTop};
        background-color: ${theme.colors.white};
    `}
`

export const ButtonWrapper = styled.div`
    display: flex;
    justify-content: space-around;
`
