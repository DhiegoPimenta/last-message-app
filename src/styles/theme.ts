export default {
    grid: {
        container: '130rem',
        gutter: '3.2rem',
    },
    border: {
        radius: '1rem',
    },
    font: {
        family: `Poppins, Ubuntu, -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, Oxygen, Cantarell, 'Open Sans', 'Helvetica Neue', sans-serif`,
        light: 300,
        normal: 400,
        bold: 600,
        extrabold: 800,
        sizes: {
            xsmall: '1.4rem',
            small: '1.8rem',
            medium: '2.4rem',
            large: '3.2rem',
        },
    },
    colors: {
        primary: '#64CBC5',
        secondary: '#0030FF',
        terciary: '#A0DFDC',
        white: '#FFFFFF',
        gray: '#CECECE',
        danger: '#FBA9B0',
        black: '#444444',
        overlay: 'rgba(0, 0, 0, 0.3)',
    },
    spacings: {
        xxsmall: '0.8rem',
        xsmall: '1.6rem',
        small: '2.4rem',
        medium: '3.2rem',
        large: '4.0rem',
        xlarge: '4.8rem',
        xxlarge: '5.6rem',
    },
    layers: {
        base: 10,
        menu: 20,
        overlay: 30,
        modal: 40,
        alwaysOnTop: 50,
    },
    transition: {
        default: '0.3s ease-in-out',
        fast: '0.1s ease-in-out',
    },
} as const
