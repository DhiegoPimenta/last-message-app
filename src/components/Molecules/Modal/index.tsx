import Button, { ButtonColorStyle } from 'components/Atoms/Button'
import Text from 'components/Atoms/Text'
import Title from 'components/Atoms/Title'

import * as S from './styles'

export type ButtonProps = {
    label: string
    handleClick: () => void
    type: ButtonColorStyle
}

export type ModalProps = {
    title: string
    description: string
    showModal?: boolean
    children?: React.ReactNode
    buttonAction?: ButtonProps
    buttonCancel?: ButtonProps
}

const Modal = ({
    children,
    title,
    description,
    buttonAction,
    buttonCancel,
}: ModalProps) => (
    <S.Overlay>
        <S.Wrapper>
            <Title size="normal" color="primary" weight="extrabold" fullWidth>
                {title}
            </Title>
            <Text tagName="p" fullWidth>
                {description}
            </Text>

            {children}

            <S.ButtonWrapper>
                {buttonCancel && (
                    <Button
                        colorStyle={buttonCancel.type}
                        onClick={() => buttonCancel.handleClick()}
                    >
                        {buttonCancel.label}
                    </Button>
                )}
                {buttonAction && (
                    <Button
                        colorStyle={buttonAction.type}
                        onClick={() => buttonAction.handleClick()}
                    >
                        {buttonAction.label}
                    </Button>
                )}
            </S.ButtonWrapper>
        </S.Wrapper>
    </S.Overlay>
)

export default Modal
