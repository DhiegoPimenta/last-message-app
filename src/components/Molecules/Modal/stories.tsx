import { Story, Meta } from '@storybook/react/types-6-0'

import Modal, { ModalProps } from '.'

export default {
    title: 'Molecules/Modal',
    component: Modal,
} as Meta

export const Basic: Story<ModalProps> = (args) => <Modal {...args} />
Basic.args = {
    title: 'title',
    description: 'description',
    buttonAction: {
        label: 'success',
        handleClick: () => {},
        type: 'info',
    },
    buttonCancel: {
        label: 'error',
        handleClick: () => {},
        type: 'danger',
    },
}
