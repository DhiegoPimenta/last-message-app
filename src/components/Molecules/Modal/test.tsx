import { fireEvent, screen } from '@testing-library/react'
import theme from 'styles/theme'
import { renderWithTheme } from 'utils/tests/helpers'

import Modal, { ButtonProps } from '.'

describe('<Modal />', () => {
    it('should render the modal', () => {
        const { container } = renderWithTheme(
            <Modal title="any-title" description="any-description">
                Modal
            </Modal>
        )

        expect(screen.getByText(/any-title/i)).toBeInTheDocument()
        expect(screen.getByText(/any-description/i)).toBeInTheDocument()
        expect(screen.getByText(/modal/i)).toBeInTheDocument()

        expect(container.firstChild).toMatchSnapshot()
    })

    it('should call funtion on click success button', () => {
        const mockHandle = jest.fn()
        const buttonAction: ButtonProps = {
            label: 'any-success',
            handleClick: mockHandle,
            type: 'success',
        }

        renderWithTheme(
            <Modal
                title="any-title"
                description="any-description"
                buttonAction={buttonAction}
            >
                Modal
            </Modal>
        )

        fireEvent.click(screen.getByText(/any-success/i))

        expect(screen.getByText(/any-success/i)).toBeInTheDocument()
        expect(screen.getByText(/any-success/i)).toHaveStyle({
            backgroundColor: theme.colors.primary,
        })

        expect(mockHandle).toBeCalled()
    })

    it('should call funtion on click error button', () => {
        const mockHandle = jest.fn()
        const buttonCancel: ButtonProps = {
            label: 'any-error',
            handleClick: mockHandle,
            type: 'danger',
        }

        renderWithTheme(
            <Modal
                title="any-title"
                description="any-description"
                buttonCancel={buttonCancel}
            >
                Modal
            </Modal>
        )

        fireEvent.click(screen.getByText(/any-error/i))

        expect(screen.getByText(/any-error/i)).toBeInTheDocument()
        expect(screen.getByText(/any-error/i)).toHaveStyle({
            backgroundColor: theme.colors.danger,
        })

        expect(mockHandle).toBeCalled()
    })
})
