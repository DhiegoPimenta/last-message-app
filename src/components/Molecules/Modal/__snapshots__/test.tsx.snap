// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal /> should render the modal 1`] = `
.c3 {
  color: #64CBC5;
  font-size: 2.4rem;
  width: 100%;
  display: block;
  font-weight: 400;
  text-align: center;
}

.c2 {
  font-weight: bold;
  color: #64CBC5;
  font-size: 2.4rem;
  font-weight: 400;
  font-weight: 800;
  width: 100%;
  text-align: center;
}

.c0 {
  background: rgba(0,0,0,0.3);
  width: 100vw;
  height: 100vh;
  position: absolute;
  z-index: 10;
  left: 0;
  top: 0;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c1 {
  width: 45rem;
  height: 30rem;
  border: 4px solid #FBA9B0;
  border-radius: 1rem;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
  z-index: 50;
  background-color: #FFFFFF;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: space-around;
  -webkit-justify-content: space-around;
  -ms-flex-pack: space-around;
  justify-content: space-around;
}

<div
  class="c0"
>
  <div
    class="c1"
  >
    <h1
      class="c2"
      color="primary"
    >
      any-title
    </h1>
    <p
      class="c3"
      color="primary"
    >
      any-description
    </p>
    Modal
    <div
      class="c4"
    />
  </div>
</div>
`;
